src/shared/sleep-config.c

```diff
@@ -184,12 +177,145 @@ static SwapEntry* swap_entry_free(SwapEntry *se) {
         return mfree(se);
 }
 
-DEFINE_TRIVIAL_CLEANUP_FUNC(SwapEntry*, swap_entry_free);
+HibernateLocation* hibernate_location_free(HibernateLocation *hl) {
+        if (!hl)
+                return NULL;
+
+        swap_entry_free(hl->swap);
+        free(hl->resume);
+
+        return mfree(hl);
+}
+
+static int swap_device_to_major_minor(const SwapEntry *swap, char **ret) {
+        _cleanup_free_ char *major_minor = NULL;
+        _cleanup_close_ int fd = -1;
+        struct stat sb;
+        dev_t swap_dev;
+        int r;
+
+        assert(swap);
+        assert(swap->device);
+        assert(swap->type);
+
+        fd = open(swap->device, O_RDONLY | O_CLOEXEC | O_NONBLOCK);
+        if (fd < 0)
+                return log_debug_errno(errno, "Unable to open '%s': %m", swap->device);
+
+        r = fstat(fd, &sb);
+        if (r < 0)
+                return log_debug_errno(errno, "Unable to stat %s: %m", swap->device);
+
+        swap_dev = streq(swap->type, "partition") ? sb.st_rdev : sb.st_dev;
+        if (asprintf(&major_minor, "%u:%u", major(swap_dev), minor(swap_dev)) < 0)
+                return log_oom();
+
+        *ret = TAKE_PTR(major_minor);
+
+        return 0;
+}
+
+static int calculate_swap_file_offset(const SwapEntry *swap, uint64_t *ret_offset) {
+        _cleanup_close_ int fd = -1;
+        _cleanup_free_ struct fiemap *fiemap = NULL;
+        struct stat sb;
+        int r, btrfs;
+
+        assert(swap);
+        assert(swap->device);
+
+        if (!streq(swap->type, "file")) {
+                log_warning("Attempted to calculate offset on swap entry that is not of type 'file': '%s'", swap->type);
+                *ret_offset = 0;
+                return 0;
+        }
+
+        fd = open(swap->device, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+        if (!fd)
+                return log_error_errno(errno, "Failed to open %s: %m", swap->device);
+
+        r = fstat(fd, &sb);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to stat %s: %m", swap->device);
+
+        btrfs = btrfs_is_filesystem(fd);
+        if (btrfs < 0)
+                return log_error_errno(r, "Error checking %s for Btrfs filesystem: %m", swap->device);
+        else if (btrfs > 0) {
+                log_debug("Detection of swap file offset on Btrfs is not supported: %s; skipping", swap->device);
```

Comment:
So here we have `return 0` and nothing is set. The general rule is: non-negative return == success == all outputs are set. Please either return negative here or set the output variable.